crates/core/src/db/datastore/locking_tx_datastore/state_view.rs

```diff
@@ -287,37 +284,33 @@ impl<'a> Iterator for Iter<'a> {
                 ScanStage::Committed { iter } => {
                     // (1) Go through the committed state for this table.
                     let _span = tracing::debug_span!("ScanStage::Committed").entered();
-                    for (row_id, row) in iter {
+                    for row_ref in iter {
                         // Increment metric for number of committed rows scanned.
                         self.num_committed_rows_fetched += 1;
-                        match self.tx_state {
-                            Some(tx_state) => {
-                                // Check the committed row's state in the current tx.
-                                match tx_state.get_row_op(&table_id, row_id) {
-                                    RowState::Committed(_) => unreachable!("a row cannot be committed in a tx state"),
-                                    // Do nothing, via (3), we'll get it in the next stage (2).
-                                    RowState::Insert(_) |
-                                    // Skip it, it's been deleted.
-                                    RowState::Delete => {}
-                                    // There either are no state changes for the current tx (`None`),
-                                    // or there are, but `row_id` specifically has not been changed.
-                                    // Either way, the row is in the committed state
-                                    // and hasn't been removed in the current tx,
-                                    // so it exists and can be returned.
-                                    RowState::Absent => return Some(DataRef::new(row_id, row)),
-                                }
-                            }
-                            None => return Some(DataRef::new(row_id, row)),
+                        // Check the committed row's state in the current tx.
+                        // If it's been deleted, skip it.
+                        // If it's still present, yield it.
+                        if !self
```

Comment:
This appears to be incorrect. We want to skip it if it's been inserted or deleted, and we only want to return it if it's not done either. See old code:

>                                   // Do nothing, via (3), we'll get it in the next stage (2).
>                                   RowState::Insert(_) |
>                                   // Skip it, it's been deleted.
 >                                   RowState::Delete => {}

Reply:
I think we should also add a test case for this.

Reply:
The previous implementation did something silly when you deleted and re-inserted a row within a transaction: it had that row present in all three of the committed state, the delete table and the insert table. The new implementation is more principled; a row can be in only the following states:

- Not present anywhere.
- Unmodified; in committed state only.
- Newly inserted; in insert table only.
- Newly deleted; in committed state and delete table.

This means that the committed state and the insert table are disjoint sets.

Reply:
Added a test in `relational_db.rs`, `test_insert_delete_insert_iter`, which exercises this path.